Add sierra gas resource tracking

THenry14 opened the pull request:
Closes #2977

This PR stack aims to add support for new resource tracking (sierra-gas) and new gas representation (GasVector triplet instead of single l1 gas value).

--> (This PR) Add sierra gas tracking
-- (https://github.com/foundry-rs/starknet-foundry/pull/3050) Run already existing tests with CairoSteps set explicitly
-- (https://github.com/foundry-rs/starknet-foundry/pull/3051) Add support for sierra gas in `--detailed-resources`
-- (https://github.com/foundry-rs/starknet-foundry/pull/3052) Add support for sierra gas in `--save-trace-data`
-- (https://github.com/foundry-rs/starknet-foundry/pull/3053) Add sierra version assertion
-- (https://github.com/foundry-rs/starknet-foundry/pull/3054) Introduce new gas representation
-- (https://github.com/foundry-rs/starknet-foundry/pull/3055) Add gas tests for sierra-gas tracking
-- (https://github.com/foundry-rs/starknet-foundry/pull/3056) Update docs with sierra-gas related changes

THenry14 committed:
Add sierra gas resource tracking

THenry14 committed:
[skip ci] Pin cairo and smoll_str versions, rebase

ddoktorski commented on crates/cheatnet/src/runtime_extensions/forge_runtime_extension/mod.rs:
```diff
@@ -730,17 +788,43 @@ pub fn get_all_used_resources(
     }
 }
+
+fn n_steps_to_sierra_gas(n_steps: usize, versioned_constants: &VersionedConstants) -> GasAmount {
+    let n_steps_u64 = u64_from_usize(n_steps);
+    let gas_per_step = versioned_constants
+        .os_constants
+        .gas_costs
+        .base
+        .step_gas_cost;
+    let n_steps_gas_cost = n_steps_u64.checked_mul(gas_per_step).unwrap_or_else(|| {
+        panic!(
+            "Multiplication overflow while converting steps to gas. steps: {n_steps}, gas per step: {gas_per_step}."
+        )
+    });
+    GasAmount(n_steps_gas_cost)
+}
+
+#[must_use]
+pub fn vm_resources_to_sierra_gas(
+    resources: &ExecutionResources,
+    versioned_constants: &VersionedConstants,
+) -> GasAmount {
+    let builtins_gas_cost =
+        builtins_to_sierra_gas(&resources.prover_builtins(), versioned_constants);
+    let n_steps_gas_cost = n_steps_to_sierra_gas(resources.total_n_steps(), versioned_constants);
+    n_steps_gas_cost.checked_add(builtins_gas_cost).unwrap_or_else(|| {
+        panic!(
+            "Addition overflow while converting vm resources to gas. steps gas: {n_steps_gas_cost}, \
+            builtins gas: {builtins_gas_cost}."
+        )
+    })
+}
```
Is Sierra gas calculated this way exactly the same as real Sierra gas calculated when executing with Cairo-native?

ddoktorski replied:
My question was not related to the conversion logic. I was wondering because what we generally do is not exactly tracking Sierra resources, but rather tracking VM resources and then converting them to Sierra resources. So my question is, are we sure that executing the same transaction with `cairo-native` will result in exactly the same Sierra gas as calculated in `snforge` with the current logic? According to the [starknet documentation](https://docs.starknet.io/architecture-and-concepts/network-architecture/fee-mechanism/#vm_resources_vs_sierra_gas), it seems that this may not necessarily be the case. Please correct me if I am wrong.

THenry14 replied:
This is not really true - we are tracking sierra resources, but the execution comes through the vm (because at the moment, we do not support cairo native execution in forge; I believe it is not even yet planned). We are converting things here, because vm has some pre-defined "spin-up" cost that is given in steps and builtins - which are not the way to describe execution cost when tracking sierra gas; when going through native the same cost would be described in sierra gas.

Regarding the docs you've linked (and answering the question) - we are as sure as we can be running thrugh vm. The cost when tracking sierra gas should not differ regardless of the environment, but we'll never be 100% sure as long as we do not support native, I guess. To reiterate my point from the previous paragraph - we are not tracking cairo steps and converting them; we are tracking sierra gas, and take them as they are reported by the vm, no conversion there. The docs state that there will be some mismatch because of the cost of the resources and max vs sum approach - which is true and you can actually observe it in the tests I added in #3055, where identical tests yield different costs depending on the tracked resource.

Hope that answers your question, but if something still seems off please let me know 🙏

ksew1 commented on crates/cheatnet/src/runtime_extensions/forge_runtime_extension/mod.rs:
```diff
@@ -661,6 +684,19 @@ fn add_syscall_resources(
+fn add_sierra_gas_resources(top_call: &Rc<RefCell<CallTrace>>) -> u64 {
+    let mut gas_consumed = top_call.borrow().gas_consumed;
+    for nested_call in &top_call.borrow().nested_calls {
+        match nested_call {
```
[nit] maybe using `if let` instead of match:


```rs
        if let CallTraceNode::EntryPointCall(nested_call) = nested_call {
            gas_consumed += &add_sierra_gas_resources(nested_call);
        }
```

ksew1 commented on crates/cheatnet/src/runtime_extensions/forge_runtime_extension/mod.rs:
```diff
@@ -730,17 +788,43 @@ pub fn get_all_used_resources(
+        .gas_costs
+        .base
+        .step_gas_cost;
+    let n_steps_gas_cost = n_steps_u64.checked_mul(gas_per_step).unwrap_or_else(|| {
```
[nit] If we are sure it is not possible, I would use `unreachable!`; otherwise, just `expect`. Unless you want to format lazily.

ksew1 commented on crates/forge-runner/src/forge_config.rs:
```diff
@@ -37,6 +43,37 @@ pub struct ExecutionDataToSave {
+    SierraGas,
+}
+
+impl FromStr for ForgeTrackedResource {
```
I can't find any use of this trait. I am missing it somewhere?

THenry14 replied:
My bad, this should be removed before the review. sorry 🙏

THenry14 committed:
Review changes

THenry14 commented on crates/forge-runner/src/forge_config.rs:
```diff
@@ -37,6 +43,37 @@ pub struct ExecutionDataToSave {
+        }
+    }
+}
+
```
I opted to add this here, as i did not want to do a circular import (forge already imports forge-runner, but forge-runner does not import forge), and these actually are a part of forge config defined in this file. 

That being said I am not super happy about it either, just couldn't find a better place for it tbh.

ddoktorski replied:
I see, probably the only option is to have another struct specifically for the config (while the current one would be only for the cli) which is not ideal either. In that case I guess it can stay as it is.

THenry14 commented on crates/cheatnet/src/runtime_extensions/forge_runtime_extension/mod.rs:
```diff
@@ -549,9 +554,11 @@ fn handle_declare_deploy_result<T: CairoSerialize>(
     runtime: &mut ForgeRuntime,
     resources: &ExecutionResources,
+    tracked_resource: &TrackedResource,
+    versioned_constants: &VersionedConstants,
```
hmm, it makes sense for versioned_constants probably 🤔 But I'd still pass tracked resource, as it is already being defined in outer scope to this function (I mean it is already defined in the only place this function is being called), so I don't really see any benefit from fetching it once again. Is that be acceptable (done in newest commit)

THenry14 commented on crates/cheatnet/src/runtime_extensions/call_to_blockifier_runtime_extension/execution/entry_point.rs:
```diff
@@ -177,13 +187,33 @@ fn remove_syscall_resources_and_exit_success_call(
-    resources -= &versioned_constants.get_additional_os_syscall_resources(syscall_counter);
+    let mut gas_consumed = call_info.execution.gas_consumed;
+
+    match &context
+        .tracked_resource_stack
+        .last()
+        .expect("Unexpected empty tracked resource.")
```
Quite the contrast from this comment by Darek https://github.com/foundry-rs/starknet-foundry/pull/3049#discussion_r1985441465 :D 

My rationale for doing this here was that we already have a context, so there was really no need to do it this way, but I'm fine with either one.

THenry14 commented on crates/cheatnet/src/runtime_extensions/forge_runtime_extension/mod.rs:
```diff
@@ -730,17 +788,43 @@ pub fn get_all_used_resources(
     }
 }
+
+fn n_steps_to_sierra_gas(n_steps: usize, versioned_constants: &VersionedConstants) -> GasAmount {
+    let n_steps_u64 = u64_from_usize(n_steps);
+    let gas_per_step = versioned_constants
+        .os_constants
+        .gas_costs
+        .base
+        .step_gas_cost;
+    let n_steps_gas_cost = n_steps_u64.checked_mul(gas_per_step).unwrap_or_else(|| {
+        panic!(
+            "Multiplication overflow while converting steps to gas. steps: {n_steps}, gas per step: {gas_per_step}."
+        )
+    });
+    GasAmount(n_steps_gas_cost)
+}
```
There is, but it's private :( And I couldn't be arsed to try and upstream making it public, sorry :(

THenry14 commented on crates/cheatnet/src/runtime_extensions/forge_runtime_extension/mod.rs:
```diff
@@ -730,17 +788,43 @@ pub fn get_all_used_resources(
+}
+
+#[must_use]
+pub fn vm_resources_to_sierra_gas(
```
Partially, but yes. Have added relevant comment and link 🙏

THenry14 commented on crates/cheatnet/src/runtime_extensions/forge_runtime_extension/mod.rs:
```diff
@@ -730,17 +788,43 @@ pub fn get_all_used_resources(
 }
+
+fn n_steps_to_sierra_gas(n_steps: usize, versioned_constants: &VersionedConstants) -> GasAmount {
+    let n_steps_u64 = u64_from_usize(n_steps);
```
This does exactly what you said it should. But hey have a helper for this in blockifier (and use it), so I figured - what the heck I'll use it too. If you rather not, I can change it to regular expect(). :+1:

THenry14 commented on crates/cheatnet/src/runtime_extensions/forge_runtime_extension/mod.rs:
```diff
@@ -716,9 +754,29 @@ pub fn get_all_used_resources(
+            );
+        }
+        TrackedResource::SierraGas => {
+            let syscalls_consumed_gas: u64 = top_call_syscalls
+                .iter()
+                .map(|(name, count)| {
+                    versioned_constants.get_syscall_gas_cost(name) * (*count as u64)
+                })
+                .sum();
+            sierra_gas_consumed += syscalls_consumed_gas;
```
To be sure - You only mean the part of fetching consumed gas by syscalls, right? 
I can do it if you think this will bring us some gains, but personaly I find it too trivial and not really worth it to make a separate function out of 🤔

cptartur replied:
This whole logic. It's not necessary, but we are kind of repeating ourselves with this.

```suggestion
            let syscalls_consumed_gas: u64 = top_call_syscalls
                .iter()
                .map(|(name, count)| {
                    versioned_constants.get_syscall_gas_cost(name) * (*count as u64)
                })
                .sum();
            sierra_gas_consumed += syscalls_consumed_gas;
```

THenry14 commented on crates/forge-runner/src/forge_config.rs:
```diff
@@ -37,6 +43,37 @@ pub struct ExecutionDataToSave {
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s.to_lowercase().as_str() {
+            "cairosteps" => Ok(Self::CairoSteps),
+            "sierragas" => Ok(Self::SierraGas),
```
No, input uses dashes. Sorry this is a leftover after some tests I forgot to remove, my bad :(

THenry14 commented on crates/forge-runner/src/running.rs:
```diff
@@ -181,7 +182,23 @@ pub fn run_test_case(
+        .versioned_constants()
+        .clone();
+
+    #[expect(clippy::clone_on_copy)]
```
Redesigned as per https://github.com/foundry-rs/starknet-foundry/pull/3049#discussion_r1985441465 and https://github.com/foundry-rs/starknet-foundry/pull/3049#discussion_r1987004084, so we no longer do it :+1:

THenry14 commented on crates/forge-runner/src/running/config_run.rs:
```diff
@@ -83,7 +87,7 @@ pub fn run_config_pass(
     let hints_dict = hints_to_params(&assembled_program);
 
-    let mut context = build_context(&block_info, None);
+    let mut context = build_context(&block_info, None, &TrackedResource::from(tracked_resource));
```
Sorry, I'm not sure I follow :( Can you rephrase or specify please?

ddoktorski replied:
This part is specifically for the so called 'configuration run' which doesn't execute any tests and only collect configuration from attributes. Tracking resources here does not matter as they are not counted for anything.

THenry14 replied:
That is true :+1: But I still don't know if any/what action is needed from me here :D

cptartur replied:
What I meant is that you probably didn't need to provide correct `tracked_resource` here and could just hardcode some value. But no need to change that.

cptartur reviewed (approved):
Looks good, just please address the comments if needed

THenry14 committed:
Review changes

THenry14 committed:
Run forge tests with cairo steps explicitly set in tests (#3050)

Related to #2977

This PR stack aims to add support for new resource tracking (sierra-gas)
and new gas representation (GasVector triplet instead of single l1 gas
value).

-- (https://github.com/foundry-rs/starknet-foundry/pull/3049) Add sierra
gas tracking
--> (This PR) Run already existing tests with CairoSteps set explicitly
-- (https://github.com/foundry-rs/starknet-foundry/pull/3051) Add
support for sierra gas in `--detailed-resources`
-- (https://github.com/foundry-rs/starknet-foundry/pull/3052) Add
support for sierra gas in `--save-trace-data`
-- (https://github.com/foundry-rs/starknet-foundry/pull/3053) Add sierra
version assertion
-- (https://github.com/foundry-rs/starknet-foundry/pull/3054) Introduce
new gas representation
-- (https://github.com/foundry-rs/starknet-foundry/pull/3055) Add gas
tests for sierra-gas tracking
-- (https://github.com/foundry-rs/starknet-foundry/pull/3056) Update
docs with sierra-gas related changes

THenry14 committed:
Fix test

piotmag769 commented on crates/forge/src/warn.rs:
```diff
@@ -14,7 +14,10 @@ pub(crate) fn warn_if_available_gas_used_with_incompatible_scarb_version(
     for test_target in test_targets {
         for case in &test_target.test_cases {
-            if case.config.available_gas == Some(0)
+            if case
+                .config
+                .available_gas
+                .as_ref().is_some_and(cheatnet::runtime_extensions::forge_config_extension::config::RawAvailableGasConfig::is_zero)
                 && ScarbCommand::version().run()?.scarb <= Version::new(2, 4, 3)
```
Can be removed tbh, we don't support 2.4.3 anymore xD
